scwx-qt/source/scwx/qt/ui/edit_marker_dialog.cpp

```diff
@@ -0,0 +1,311 @@
+   bool            adding_ {false};
+   std::string     setIconOnAdded_ {""};
+};
+
+QIcon EditMarkerDialog::Impl::get_colored_icon(
+   const types::MarkerIconInfo& marker, const std::string& color)
+{
+   return util::modulateColors(marker.qIcon,
+                               self_->ui->iconComboBox->iconSize(),
+                               QColor(QString::fromStdString(color)));
+}
+
+EditMarkerDialog::EditMarkerDialog(QWidget* parent) :
+    QDialog(parent),
+    p {std::make_unique<Impl>(this)},
+    ui(new Ui::EditMarkerDialog)
+{
+   ui->setupUi(this);
+
+   for (auto& markerIcon : p->markerManager_->get_icons())
+   {
+      ui->iconComboBox->addItem(markerIcon.second.qIcon,
+                                QString(""),
+                                QString::fromStdString(markerIcon.second.name));
+   }
+   p->deleteButton_ =
+      ui->buttonBox->addButton("Delete", QDialogButtonBox::DestructiveRole);
+   p->connect_signals();
+}
+
+EditMarkerDialog::~EditMarkerDialog()
+{
+   delete ui;
+}
+
+void EditMarkerDialog::setup()
+{
+   setup(0, 0);
+}
+
+void EditMarkerDialog::setup(double latitude, double longitude)
+{
+   // By default use foreground color as marker color, mainly so the icons
+   // are vissable in the dropdown menu.
+   const QColor color = QWidget::palette().color(QWidget::foregroundRole());
+   p->editId_         = p->markerManager_->add_marker(types::MarkerInfo(
+      "",
+      latitude,
+      longitude,
+      manager::MarkerManager::getDefaultIconName(),
+      boost::gil::rgba8_pixel_t {static_cast<uint8_t>(color.red()),
+                                 static_cast<uint8_t>(color.green()),
+                                 static_cast<uint8_t>(color.blue()),
+                                 static_cast<uint8_t>(color.alpha())}));
+
+   setup(p->editId_);
+   p->adding_ = true;
+}
+
+void EditMarkerDialog::setup(types::MarkerId id)
+{
+   std::optional<types::MarkerInfo> marker = p->markerManager_->get_marker(id);
+   if (!marker)
+   {
+      return;
+   }
+
+   p->editId_ = id;
+   p->adding_ = false;
+
+   const std::string iconColorStr =
+      util::color::ToArgbString(marker->iconColor);
+   p->set_icon_color(iconColorStr);
+
+   int iconIndex =
+      ui->iconComboBox->findData(QString::fromStdString(marker->iconName));
+   if (iconIndex < 0 || marker->iconName == "")
+   {
+      iconIndex = 0;
+   }
+
+   ui->nameLineEdit->setText(QString::fromStdString(marker->name));
+   ui->iconComboBox->setCurrentIndex(iconIndex);
+   ui->latitudeDoubleSpinBox->setValue(marker->latitude);
+   ui->longitudeDoubleSpinBox->setValue(marker->longitude);
+   ui->iconColorLineEdit->setText(QString::fromStdString(iconColorStr));
+}
+
+types::MarkerInfo EditMarkerDialog::get_marker_info() const
+{
+   const QString                   colorName = ui->iconColorLineEdit->text();
+   const boost::gil::rgba8_pixel_t color =
+      util::color::ToRgba8PixelT(colorName.toStdString());
+
+   return types::MarkerInfo(
```

Comment:
If you think that braced initializer is better, I am fine with using it, but I dislike it. I find it can make reading the code harder (1) and is sometimes more error prone (2).

1. You need to remember what the return type is. It is also just another piece of syntax and control flow you have to remember, but that might just be a me problem.
2. If the return type of the function is changed but not all of the return statements are checked, it could end up returning an odd result.